File: packages/moss-repo-indexer/src/moss_repo_indexer/chunkers/markdown.py
"""Heading-aware Markdown chunker."""

from __future__ import annotations

import re

from moss import DocumentInfo

from ..types import CHUNK_TYPE_HEADER, CHUNK_TYPE_PAGE, CHUNK_TYPE_TEXT
from .common import ChunkSlice, FileChunkRequest, make_document, relative_path, split_lines

_HEADING_RE = re.compile(r"^(#{1,6})\s+(.+?)\s*$")


def chunk_markdown(request: FileChunkRequest) -> list[DocumentInfo]:
    lines = split_lines(request.content)
    docs = [_page_document(request)]
    sections = _parse_sections(lines, relative_path(request.path, request.root))
    for section in sections:
        docs.extend(_documents_for_section(request, section))
    return docs


def _page_document(request: FileChunkRequest) -> DocumentInfo:
    rel = relative_path(request.path, request.root)
    title = request.path.stem
    end_line = max(1, len(split_lines(request.content)))
    generated = make_document(
        request,
        ChunkSlice(
            start_line=1,
            end_line=end_line,
            body=title,
            chunk_type=CHUNK_TYPE_PAGE,
            symbol="",
            title=title,
        ),
    )
    return DocumentInfo(id=rel, text=generated.text, metadata=generated.metadata)


def _parse_sections(
    lines: list[str],
    fallback_title: str,
) -> list[tuple[str, int, int, str]]:
    """Return (heading, start_line, end_line, body) tuples."""
    sections: list[tuple[str, int, list[str]]] = []
    heading = fallback_title
    start = 1
    body: list[str] = []

    for index, line in enumerate(lines, start=1):
        match = _HEADING_RE.match(line)
        if match:
            _flush_section(sections, heading, start, body)
            heading = match.group(2).strip()
            start = index
            body = [line]
            continue
        body.append(line)

    _flush_section(sections, heading, start, body)
    return [(h, s, s + len(b) - 1, "\n".join(b).strip()) for h, s, b in sections if b]


def _flush_section(
    sections: list[tuple[str, int, list[str]]],
    heading: str,
    start: int,
    body: list[str],
) -> None:
    if body:
        sections.append((heading, start, list(body)))


def _documents_for_section(
    request: FileChunkRequest,
    section: tuple[str, int, int, str],
) -> list[DocumentInfo]:
    heading, start_line, end_line, body = section
    docs = [
        make_document(
            request,
            ChunkSlice(
                start_line=start_line,
                end_line=start_line,
                body=heading,
                chunk_type=CHUNK_TYPE_HEADER,
                symbol=heading,
                title=heading,
            ),
        )
    ]
    body_without_heading = _strip_leading_heading(body)
    if body_without_heading:
        docs.append(
            make_document(
                request,
                ChunkSlice(
                    start_line=start_line,
                    end_line=end_line,
                    body=body_without_heading,
                    chunk_type=CHUNK_TYPE_TEXT,
                    symbol=heading,
                    title=heading,
                ),
            )
        )
    return docs


def _strip_leading_heading(body: str) -> str:
    lines = body.splitlines()
    if lines and _HEADING_RE.match(lines[0]):
        return "\n".join(lines[1:]).strip()
    return body.strip()

File: packages/moss-repo-indexer-js/src/chunkers/markdown.ts
import path from 'node:path'

import type { RepoDocument } from '../document.js'

import { CHUNK_TYPE_HEADER, CHUNK_TYPE_PAGE, CHUNK_TYPE_TEXT } from '../types.js'
import { FileChunkRequest, makeDocument, relativePath, splitLines } from './common.js'

const HEADING_RE = /^(#{1,6})\s+(.+?)\s*$/

export function chunkMarkdown(request: FileChunkRequest): RepoDocument[] {
  const lines = splitLines(request.content)
  const docs: RepoDocument[] = [pageDocument(request)]
  const sections = parseSections(lines, relativePath(request.path, request.root))
  for (const section of sections) {
    docs.push(...documentsForSection(request, section))
  }
  return docs
}

function pageDocument(request: FileChunkRequest): RepoDocument {
  const rel = relativePath(request.path, request.root)
  const title = path.basename(request.path, path.extname(request.path))
  const endLine = Math.max(1, splitLines(request.content).length)
  const generated = makeDocument(request, {
    startLine: 1,
    endLine,
    body: title,
    chunkType: CHUNK_TYPE_PAGE,
    symbol: '',
    title,
  })
  return { id: rel, text: generated.text, metadata: generated.metadata }
}

type Section = { heading: string; startLine: number; endLine: number; body: string }

function parseSections(lines: string[], fallbackTitle: string): Section[] {
  const buffered: Array<{ heading: string; start: number; body: string[] }> = []
  let heading = fallbackTitle
  let start = 1
  let body: string[] = []

  lines.forEach((line, index) => {
    const lineNo = index + 1
    const match = HEADING_RE.exec(line)
    if (match) {
      flushSection(buffered, heading, start, body)
      heading = match[2].trim()
      start = lineNo
      body = [line]
      return
    }
    body.push(line)
  })
  flushSection(buffered, heading, start, body)

  return buffered
    .filter((item) => item.body.length > 0)
    .map((item) => ({
      heading: item.heading,
      startLine: item.start,
      endLine: item.start + item.body.length - 1,
      body: item.body.join('\n').trim(),
    }))
}

function flushSection(
  sections: Array<{ heading: string; start: number; body: string[] }>,
  heading: string,
  start: number,
  body: string[],
): void {
  if (body.length > 0) {
    sections.push({ heading, start, body: [...body] })
  }
}

function documentsForSection(request: FileChunkRequest, section: Section): RepoDocument[] {
  const docs: RepoDocument[] = [
    makeDocument(request, {
      startLine: section.startLine,
      endLine: section.startLine,
      body: section.heading,
      chunkType: CHUNK_TYPE_HEADER,
      symbol: section.heading,
      title: section.heading,
    }),
  ]
  const bodyWithoutHeading = stripLeadingHeading(section.body)
  if (bodyWithoutHeading) {
    docs.push(
      makeDocument(request, {
        startLine: section.startLine,
        endLine: section.endLine,
        body: bodyWithoutHeading,
        chunkType: CHUNK_TYPE_TEXT,
        symbol: section.heading,
        title: section.heading,
      }),
    )
  }
  return docs
}

function stripLeadingHeading(body: string): string {
  const lines = body.split(/\r?\n/)
  if (lines.length > 0 && HEADING_RE.test(lines[0])) {
    return lines.slice(1).join('\n').trim()
  }
  return body.trim()
}
